walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,126 @@
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<String> attendees = new ArrayList<String>(request.getAttendees());
+    long longDuration = request.getDuration();
+    ArrayList<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+
+    // Deal with special cases:
+    // Meeting is too long, impossible to ever hold it
+    if (longDuration > TimeRange.WHOLE_DAY.duration()) {
+      return possibleTimes;
+    }
+    // No events/no attendees, hold meeting anytime
+    if (events.size() == 0 || attendees.size() == 0) {
+      possibleTimes.add(TimeRange.fromStartDuration(TimeRange.START_OF_DAY, TimeRange.WHOLE_DAY.duration()));
+      return possibleTimes;
+    }
+ 
+    // Since we know longDuration is less than one day, it can fit in an int
+    int duration = (int)longDuration;
+
+    // Filter events so that it only includes events that have one or more
+    // attendees who are coming to this event
+    ArrayList<Event> eventsByStart = new ArrayList<Event>();
+    for (Event event : events) {
+      HashSet<String> eventAttendees = new HashSet<String>(event.getAttendees());
+      HashSet<String> meetingAttendees = new HashSet<String>(attendees);
+      eventAttendees.retainAll(meetingAttendees); // set intersection
+      if (eventAttendees.size() > 0) {
+        eventsByStart.add(event);
+      }
+    }
+    Collections.sort(eventsByStart, Event.ORDER_BY_START_TIME);   
+
+    // Check again to see if there are any events
+    if (eventsByStart.size() == 0 ) {
+      possibleTimes.add(TimeRange.fromStartDuration(TimeRange.START_OF_DAY, TimeRange.WHOLE_DAY.duration()));
+      return possibleTimes;
+    }
+
+    // Check if we can have a meeting before the earliest-starting event
+    int earliestStart = eventsByStart.get(0).getWhen().start();
+    if (TimeRange.START_OF_DAY < earliestStart && earliestStart - TimeRange.START_OF_DAY >= duration) {
+      possibleTimes.add(TimeRange.fromStartEnd(
+          TimeRange.START_OF_DAY, earliestStart, false));
+    }
+
+    // Check in between events - we know we can't have anything until
+    // the earliest-starting event ends, so find the earliest-starting event
+    // after that end time and see if the meeting fits between them
+    Event currentEarliestEvent = eventsByStart.get(0);
+    Event nextEvent;
+    while (true) {
+      // Try to get event
+      Optional<Event> potentialNextEvent = getNextEarliestEvent(
+          currentEarliestEvent.getWhen().end(), eventsByStart, 0, eventsByStart.size()-1);
+
+      if (!potentialNextEvent.isPresent()) { 
+        break; // No event starts after current one ends
+      }
+
+      // Check if we can fit the meeting between the two events
+      // ***BUG***
```

Comment:
I'm not sure exactly what the bug is but I think it may have to do with not considering in potentialNextEvent() that an event can start and end before the next event starts. ie.
|--------A--------|
&nbsp;&nbsp;&nbsp;&nbsp;|---B---|
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|---C---|

Reply:
Yeah, this is definitely the bug - I describe a similar situation in my pull request comment, and you can take a look at the test I wrote to see another such situation. Thanks for the confirmation! I will probably send my test in as a pull request to the main repo.